module/documents/actor/actor.mjs

```diff
@@ -1117,29 +1213,58 @@ export default class Actor5e extends Actor {
 
     // Prepare roll data
     const flavor = game.i18n.localize("DND5E.HitDiceRoll");
-
-    // Call the roll helper utility
-    const roll = await damageRoll({
+    if ( options.fastForward === undefined ) options.fastForward = !options.dialog;
+    const rollData = foundry.utils.mergeObject(options, {
       event: new Event("hitDie"),
       parts: [`1${denomination}`, "@abilities.con.mod"],
       data: this.toObject().system,
       title: `${flavor}: ${this.name}`,
       flavor,
       allowCritical: false,
-      fastForward: !dialog,
       dialogOptions: {width: 350},
       messageData: {
         speaker: ChatMessage.getSpeaker({actor: this}),
         "flags.dnd5e.roll": {type: "hitDie"}
       }
     });
-    if ( !roll ) return null;
 
-    // Adjust actor data
-    await cls.update({"system.hitDiceUsed": cls.system.hitDiceUsed + 1});
+    /**
+     * A hook event that fires before a hit die is rolled for an Actor.
+     * @function dnd5e.preRollHitDie
+     * @memberof hookEvents
+     * @param {Actor5e} actor                Actor for which the hit die is to be rolled.
+     * @param {D20RollConfiguration} config  Configuration data for the pending roll.
+     * @param {string} denomination          Size of hit die to be rolled.
+     * @returns {boolean}                    Explicitly return false to prevent hit die from being rolled.
+     */
+    if ( Hooks.call("dnd5e.preRollHitDie", this, rollData, denomination) === false ) return;
+
+    const roll = await damageRoll(rollData);
+    if ( !roll ) return roll;
+
     const hp = this.system.attributes.hp;
     const dhp = Math.min(hp.max + (hp.tempmax ?? 0) - hp.value, roll.total);
-    await this.update({"system.attributes.hp.value": hp.value + dhp});
+    const updates = {
+      actor: {"system.attributes.hp.value": hp.value + dhp},
+      class: {"system.hitDiceUsed": cls.system.hitDiceUsed + 1}
+    };
+
+    /**
+     * A hook event that fires after a hit die has been rolled for an Actor.
+     * @function dnd5e.rollHitDie
+     * @memberof hookEvents
+     * @param {Actor5e} actor        Actor for which the hit die has been rolled.
+     * @param {D20Roll} roll         The resulting roll.
+     * @param {object} updates
+     * @param {object} updates.actor  Updates that will be applied to the actor.
+     * @param {object} updates.class  Updates that will be applied to the class.
+     */
+    if ( Hooks.call("dnd5e.rollHitDie", this, roll, updates) === false ) return roll;
+
+    // Perform updates
+    if ( !foundry.utils.isEmpty(updates.actor) ) this.update(updates.actor);
```

Comment:
Is it worth awaiting these operations, or are we trusting that the server handles concurrent requests on embedded items/parent documents?